tools/clang/lib/Sema/SemaHLSLDiagnoseTU.cpp

```diff
@@ -325,6 +325,136 @@ class HLSLCallDiagnoseVisitor
     return true;
   }
 
+  bool VisitDeclRefExpr(DeclRefExpr *ET) {
+    // Search for enum types that should only exist after
+    // specific shader models.
+    IdentifierInfo *II = ET->getDecl()->getIdentifier();
+
+    if (!SM->IsSMAtLeast(6, 9)) {
+      if (II && II->getName().startswith("RAYQUERY_FLAG")) {
+        sema->Diag(ET->getLocation(), diag::warn_hlsl_enum_type_not_allowed)
+            << II->getName() << SM->GetName() << "6.9";
+      }
+    }
+
+    return true;
+  }
+
+  bool IsRayFlagForceOMM2StateSet(const CallExpr *CE) {
+    const ImplicitCastExpr *RayFlagArg =
+        dyn_cast<ImplicitCastExpr>(CE->getArg(1));
+    if (!RayFlagArg)
+      return false;
+
+    const DeclRefExpr *DR = dyn_cast<DeclRefExpr>(RayFlagArg->getSubExpr());
+    if (DR) {
+      const ValueDecl *valueDecl = DR->getDecl();
+      const VarDecl *varDecl = dyn_cast<VarDecl>(valueDecl);
+
+      if (varDecl) {
+        const Expr *initializer = varDecl->getInit();
+        if (!initializer)
+          return false;
+
+        const IntegerLiteral *intLit = dyn_cast<IntegerLiteral>(initializer);
+        if (!intLit)
+          return false;
+
+        unsigned int rayFlagValue = intLit->getValue().getZExtValue();
+        if (rayFlagValue & (unsigned int)DXIL::RayFlag::ForceOMM2State) {
+          return true;
+        }
+      }
+    } else if (const IntegerLiteral *intLit =
+                   dyn_cast<IntegerLiteral>(RayFlagArg->getSubExpr())) {
+      unsigned int rayFlagValue = intLit->getValue().getZExtValue();
+      if (rayFlagValue & (unsigned int)DXIL::RayFlag::ForceOMM2State)
+        return true;
+    }
+    return false;
+  }
+
+  void DiagnoseTraceRayInline(const MemberExpr *ME,
+                              CXXMemberCallExpr *callExpr) {
+    // validate that if the ray flag parameter is
+    // RAY_FLAG_FORCE_OMM_2_STATE, the ray query object
+    // has the RAYQUERY_FLAG_ALLOW_OPACITY_MICROMAPS
+    // flag set, otherwise emit a diagnostic
```

Comment:
Not "if the ray flag parameter is RAY_FLAG_FORCE_OMM_2_STATE", but "if the ray flag parameter has the RAY_FLAG_FORCE_OMM_2_STATE flag set".
```suggestion
    // If the RayFlag parameter has RAY_FLAG_FORCE_OMM_2_STATE set,
    // the RayQuery decl must have RAYQUERY_FLAG_ALLOW_OPACITY_MICROMAPS set,
    // otherwise emit a diagnostic.
```